Add SQL Check Operator Guide (#270)
astronomer/airflow-guides: guides/airflow-sql-data-quality-tutorial.md

---

> title: "Using Airflow to Execute SQL for Data Quality Use Cases"

Reviewer: Can we change the title to be something a little more distinct from our other SQL guide? Maybe something like "Airflow Data Quality Checks with SQL"...

description: "Executing queries, parameterizing queries, and embedding SQL-driven ETL in Apache Airflow DAGs to ensure data quality."
date: 2021-09-09T00:00:00.000Z
slug: "airflow-sql-data-quality-tutorial"

> tags: ["Database", "SQL", "DAGs"]

Reviewer: ```suggestion
tags: ["Database", "SQL", "DAGs", "Data Quality"]
```

Reviewer: Let's add a "Data Quality" tag here, I think we'll want that in the future if you add any more of these :)

---

> Note: All code in this guide can be found in [this Github repo](https://github.com/astronomer/airflow-data-quality-demo/).

## Overview

> Executing SQL queries is one of the most common use cases for data pipelines and a simple and effective way to implement data quality checks. Using Airflow, you can quickly put together a pipeline specifically for checking data quality or add quality checks to existing ETL/ELT with just a few lines of boilerplate code.

Reviewer: I think it would be helpful to add a sentence or two up front about the use case for implementing within-DAG data quality checks to set the stage before diving in to using SQL to do so.

In this guide, we'll introduce the different SQL Check Operators and show examples of how each one can be used to build a robust data quality suite for your DAGs. If you aren't familiar with SQL Operators in general, check out Astronomer's [SQL tutorial](https://www.astronomer.io/guides/airflow-sql-tutorial) first.

## SQL Check Operators

The SQL Check Operators are specific versions of the `SQLOperator` that streamline different checks against queries. One main difference between the Check Operators and the standard [`BaseSQLOperator`](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/operators/sql/index.html#airflow.operators.sql.BaseSQLOperator) is that the Check Operators ultimately respond with a boolean, meaning the task will fail if the resulting query fails. This is extremely helpful in stopping a data pipeline before bad data makes it to the destination. With Airflow's logging capabilities, knowing exactly which line (and even which value in that line) failed the check becomes highly observable.

> Airflow currently supports the following SQL Check Operators:

Reviewer: Do you think it would be helpful in this section to have a brief overview of when you would want to use each of these operators? I think you cover that within each example, but if somebody were reading this as an overview I think it would be helpful to have something to point them in the right direction without having to read each section in detail.

- [SQLCheckOperator](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/operators/sql/index.html#airflow.operators.sql.SQLCheckOperator)

- [SQLValueCheckOperator](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/operators/sql/index.html#airflow.operators.sql.SQLValueCheckOperator)

- [SQLIntervalCheckOperator](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/operators/sql/index.html#airflow.operators.sql.SQLIntervalCheckOperator)

- [SQLThresholdCheckOperator](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/operators/sql/index.html#airflow.operators.sql.SQLThresholdCheckOperator)

## Examples

> The following examples show you how to use each of the aforementioned operators and give some background on when each one will be most useful.

Reviewer: Somewhere in here I think it would be helpful to show one of the operators in a full example DAG, just to provide some context on how they would be integrated into a full pipeline. Even just a screenshot of a graph view of such a DAG with a few sentences of explanation would probably be helpful. And definitely doesn't need to be for every example, could just be the first one.

Reviewer 2: Agree with kenten here! We should look into how to get the nice graph views from the registry embedded into the guides site!

### Example 1 - `SQLCheckOperator`

> The `SQLCheckOperator` returns a single row from a provided SQL query and checks to see if any of the returned values in that row are `False`. If any values are `False`, the task fails. This operator allows a great deal of flexibility in what is being checked: a specific single value, NULL-checking and other type-checking, values of an entire row compared to a known set of values, specific calculations like `SUM`s or `AGG`s, or any other function that can be written into a SQL query.

Reviewer: I would bullet the list here, since it's pretty long

The following code snippet shows you how to use the operator in a DAG.

```python
SQLCheckOperator(
    task_id=f"forestfire_row_quality_check",

>     sql="sql/row_quality_yellow_tripdata_check.sql",

Reviewer: I typically put code in these guides in the context of an Astronomer project - so using `include/` for any additional scripts that are referenced in the DAGs. Or something like `include/sql/` if there needs to be more hierarchy. Not required, just a thought :)

Author: I've had issues with that in the past with `sql/` directories and this operator in particular. I'll try messing with it more and see if it was just some misunderstanding on my part.

Author: Realized I also needed to add a template search path to the DAG args, now it picks up files in `include/` just fine.

    params=values,
)
```

> The `sql` argument can be either the complete SQL query as a string or, as in this example, a reference to a query in a local file. The `params` argument allows you to pass a dictionary of values to the SQL query, which can be accessed through the `params` keyword in the query, as shown in the SQL code snippet below. The `database` and `conn_id` arguments are left out of the example, so the default values will be used. The full code can be found in the [data quality demo repository](https://github.com/astronomer/airflow-data-quality-demo/).

Reviewer: I think you get at this indirectly throughout, but I would be explicit about the need for correctly crafting your SQL queries when using these operators. Maybe just explaining directly how you created the query below to address the use case you describe would be enough. Mostly say this because I've seen some confusion around this in the past.

This operator works great for use cases where you know certain metrics should be close in value, or have close statistics, to the same metrics on previous days, weeks, or other time scales.

The code snippet below shows how to use this operator to check that the average distance of a taxi ride today is within 1.5 times the average distance of yesterday's rides:

```python
SQLIntervalCheckOperator(
    task_id="check_interval_data",
    days_back=1,
    date_filter_column="upload_date",
    metrics_threshold={"AVG(trip_distance)": 1.5}
)
```

In the task above, we are only looking one day back in the `days_back` argument, comparing the current day's data to yesterday's. We overwrite the `date_filter_column` to `upload_date`, which defaults to `ds`. The `metrics_threshold` is set to a dictionary of one key-value pair, where the key is a SQL expression that returns a result used in the default ratio formula, and the value of the dictionary is the threshold that the calculated ratio must not exceed.

### Example 4 - `SQLThresholdCheckOperator`

> The `SQLThresholdCheckOperator` works similarly to the `SQLValueCheckOperator`, except instead of a single threshold, there is a min and max given, making this operator more dynamic in the type and scale of thresholding. The min and max thresholds can be numeric types or SQL expressions that result in a numeric type. Only the first result in the given SQL queries, for the `sql` argument and threshold arguments, will be used in the evaluation. This operator is great for use cases where a metric like a minimum, maximum, sum, or aggregate, is known to have to be between certain values.

Reviewer: "Only the first result in the given SQL queries, for the `sql` argument and threshold arguments, will be used in the evaluation."

I'm not sure I understand what this means...are you saying if you provide multiple SQL queries only the first will be run? And how is that different from the other operators?

Author: Ugh, yeah, this is a really confusing (and poorly documented) part of these operators. Only the first row of a result set is returned by the operator. I'll clean up the language. But it is true for (iirc) all the operators, so maybe that point should be moved up to a more general section.

    - An operator with flexible upper and lower thresholds, where the threshold bounds may also be described as SQL queries that return a numeric value

## Examples

The following examples show you how to use each of the aforementioned operators and give some background on when each one will be most useful.

### Example 1 - `SQLCheckOperator`

The `SQLCheckOperator` returns a single row from a provided SQL query and checks to see if any of the returned values in that row are `False`. If any values are `False`, the task fails. This operator allows a great deal of flexibility in what is being checked:

- A specific single value
- NULL-checking and other type-checking
- Values of an entire row compared to a known set of values
- Specific calculations like `SUM`s or `AGG`s
- Any other function that can be written as a SQL query

The following code snippet shows you how to use the operator in a DAG.

```python
SQLCheckOperator(
    task_id="yellow_tripdata_row_quality_check",
    sql="row_quality_yellow_tripdata_check.sql",
    params=values,
)
```

> The `sql` argument can be either the complete SQL query as a string or, as in this example, a reference to a query in a local file (for Astronomer projects, this is in the `include/` directory). The `params` argument allows you to pass a dictionary of values to the SQL query, which can be accessed through the `params` keyword in the query, as shown in the SQL code snippet below. The `database` and `conn_id` arguments are left out of the example, so the default values will be used. The full code can be found in the [data quality demo repository](https://github.com/astronomer/airflow-data-quality-demo/).

Reviewer 3: ```suggestion
The `sql` argument can be either the complete SQL query as a string or, as in this example, a reference to a query in a local file (for Astronomer projects, this is in the `include/` directory). The `params` argument allows you to pass a dictionary of values to the SQL query, which can be accessed through the `values` keyword in the query. The `database` and `conn_id` arguments are left out of the example, so the default values will be used. The full code can be found in the [data quality demo repository](https://github.com/astronomer/airflow-data-quality-demo/).
```
Assume you meant `values` here? Don't see `params` in the query below.

The query used in the `sql` argument is:

```sql
SELECT vendor_id, pickup_datetime,
  CASE WHEN dropoff_datetime > pickup_datetime THEN 1 ELSE 0 END AS date_check,
  CASE WHEN passenger_count >= 0 THEN 1 ELSE 0 END AS passenger_count_check,
  CASE WHEN trip_distance >= 0 AND trip_distance <= 100 THEN 1 ELSE 0 END AS trip_distance_check,
  CASE WHEN ROUND((fare_amount + extra + mta_tax + tip_amount + improvement_surcharge + COALESCE(congestion_surcharge, 0)), 1) = ROUND(total_amount, 1) THEN 1
       WHEN ROUND(fare_amount + extra + mta_tax + tip_amount + improvement_surcharge, 1) = ROUND(total_amount, 1) THEN 1 ELSE 0 END AS fare_check
FROM yellow_tripdata
WHERE pickup_datetime IN (SELECT pickup_datetime FROM yellow_tripdata ORDER BY RANDOM() LIMIT 1)
```

> By using `CASE` statements in the SQL query, we can check very specific cases of data quality that we, as the data stewards, know should always be true:

Reviewer 3: ```suggestion
By using `CASE` statements in the SQL query, we can check very specific cases of data quality that should always be true for this use case:
```
Wouldn't assume that the person reading this is a data steward.

> - Drop-offs always occur after pickups

Reviewer 3: ```suggestion
- Drop-offs always occur after pickups.
```
We can punctuate this list because these items are complete sentences.

- A trip is only valid if there is at least one passenger
- A trip needs to be in a range allowed by the taxi company (in this case, we assume there is a maximum allowed trip distance of 100 miles)
- Each of the components of the total fare should add up to the total fare

> Using a for loop, tasks are generated to run this check on every row of the data or on any desired subset of the data for spot-checks. In the SQL above, a `pickup_datetime` is chosen randomly, and the corresponding code uses a loop to spot-check ten rows. In the example DAG below, we can see how the loop results in `TaskGroups` that can be collapsed or expanded in the Airflow UI.

Reviewer 3: ```suggestion
Using a for loop, tasks are generated to run this check on every row or other subset of the data. In the SQL above, a `pickup_datetime` is chosen randomly, and the corresponding code uses a loop to spot-check ten rows. In the example DAG below, we can see how the loop results in `TaskGroups` that can be collapsed or expanded in the Airflow UI:
```

> ![An example DAG showing data quality checks as part of a pipeline.](https://assets2.astronomer.io/main/guides/sql-data-quality-tutorial/example_dq_dag.png)

Reviewer 3: Love the descriptive alt text!

> In the example DAG above, we see exactly where the data quality aspect fits into a pipeline. By loading the data into Redshift then performing checks as queries, we are offloading compute resources from Airflow to Redshift, allowing Airflow to act only as an orchestrator. For a production pipeline, data could first be loaded from S3 to a temporary staging table, quality checks performed, and then another `SQLOperator` can load the data from staging to a production table. This way, if the data quality checks fail, the pipeline can be stopped, and the staging table can either be used to help diagnose the data issue or scrapped to save resources. To see the complete example DAG and run it for yourself, check out the [data quality demo repository](https://github.com/astronomer/airflow-data-quality-demo/).

Reviewer 3: ```suggestion
In the example DAG above, we see exactly how our data quality checks fit into a pipeline. By loading the data into Redshift then performing checks as queries, we are offloading compute resources from Airflow to Redshift, which frees up Airflow to act only as an orchestrator. 

For a production pipeline, data could first be loaded from S3 to a temporary staging table, then have its quality checks completed. If the quality checks succeed, another `SQLOperator` can load the data from staging to a production table. If the data quality checks fail, the pipeline can be stopped, and the staging table can be either used to help diagnose the data issue or scrapped to save resources. To see the complete example DAG and run it for yourself, check out the [data quality demo repository](https://github.com/astronomer/airflow-data-quality-demo/).
```

### Example 2 - `SQLValueCheckOperator`

> The `SQLValueCheckOperator` is simpler than the `SQLCheckOperator` but still incredibly useful. It checks the results of a query against a specific pass value, and, if given, ensures the value is within a percentage threshold. The pass value can be any type, but the threshold can only be used with numeric types. This check is useful for:

Reviewer 3: ```suggestion
The `SQLValueCheckOperator` is simpler than the `SQLCheckOperator` but equally useful. It checks the results of a query against a specific pass value, and ensures the checked value is within a percentage threshold of the pass value. The pass value can be any type, but the threshold can only be used with numeric types. This check is useful for:
```

> - Checking the row count of a table

Reviewer 3: ```suggestion
- Checking the row count of a table.
```
This list can be capitalized because it completes the sentence that the introductory clause starts.

> - Ensuring all text values in a record or set of records are correct (say, checking text country codes against numeric ones)

Reviewer 3: ```suggestion
- Ensuring all text values in a record or set of records are correct (for example, checking text country codes against numeric ones).
```

The task below shows how to use this operator to check the row count of a table within a tolerance threshold:

```python
SQLValueCheckOperator(
    task_id="check_row_count",
    sql=f"SELECT COUNT(*) FROM yellow_tripdata",
    pass_value=20000,
    threshold=0.01
)
```

The `sql` argument is a query that ensures the actual table row count is our known `pass_value`, 20000, and that the `pass_value` is within a 1% tolerance, given in the `threshold` argument. The SQL query is a simple one-liner, so we just include it as a string directly in the argument field. Again, `database` and `conn_id` are left as defaults.

### Example 3 - `SQLIntervalCheckOperator`

> The `SQLIntervalCheckOperator` is a little more complex than the previous operators. It operates on a time-based interval, which defaults to choosing a reference day seven days prior to the run date (using Airflow's `{{ ds }}` macro). A set of metric thresholds are given as a dictionary for each metric that is being checked, with the keys of the dictionary each being a SQL expression that returns data for the current day and the reference day, and the values of the dictionary each being the ratio used as a threshold. The current day's data are compared against the reference day's data using a ratio formula, and if the ratio formula produces a result outside of the threshold, the task fails. The two formulas are:

Reviewer 3: ```suggestion
The `SQLIntervalCheckOperator` is a little more complex than the previous operators. By default, it checks data against reference data from seven days prior to the run date (using Airflow's `{{ ds }}` macro). You provide a set of metric thresholds as a dictionary for each metric that is being checked. The keys of the dictionary are SQL expressions that return data for the current day and the reference day, and the values of the dictionary are the ratios used as a threshold. The current day's data is compared against the reference day's data using a ratio formula, and if the ratio formula produces a result outside of the threshold, the task fails. The two formulas are:
```

- Max-over-min (default), which calculates the maximum of the current day's and reference day's data divided by the minimum of the two: `max(cur, ref) / min(cur, ref)`
- Relative difference, which calculates the absolute value of the difference between the current and reference days, divided by the reference day: `abs(cur-ref) / ref`

> This operator works great for use cases where you know certain metrics should be close in value, or have close statistics, to the same metrics on previous days, weeks, or other time scales.

Reviewer 3: ```suggestion
This operator works well for use cases where you know certain metrics should be close in value, or have close statistics, to the same metrics on previous days, weeks, or other time scales.
```

> The code snippet below shows how to use this operator to check that the average distance of a taxi ride today is within 1.5 times the average distance of yesterday's rides:

Reviewer 3: ```suggestion
The code snippet checks to see whether the average distance of a taxi ride today is within 1.5 times the average distance of yesterday's rides:
```

```python
SQLIntervalCheckOperator(
    task_id="check_interval_data",
    days_back=1,
    date_filter_column="upload_date",
    metrics_threshold={"AVG(trip_distance)": 1.5}
)
```

> In the task above, we are only looking one day back in the `days_back` argument, comparing the current day's data to yesterday's. We overwrite the `date_filter_column` to `upload_date`, which defaults to `ds`. The `metrics_threshold` is set to a dictionary of one key-value pair, where the key is a SQL expression that returns a result used in the default ratio formula, and the value of the dictionary is the threshold that the calculated ratio must not exceed.

Reviewer 3: ```suggestion
In the task above, we are looking only one day back using the `days_back` argument. We set the `date_filter_column` argument  to `upload_date`, which defaults to `ds`. We set the `metrics_threshold` argument to a dictionary of one key-value pair, where the key is a SQL expression that returns results for both dates and the value is the threshold that the calculated ratio of the results must not exceed.
```

### Example 4 - `SQLThresholdCheckOperator`

> The `SQLThresholdCheckOperator` works similarly to the `SQLValueCheckOperator`, except instead of a single threshold, there is a min and max given, making this operator more dynamic in the type and scale of thresholding. The min and max thresholds can be numeric types or SQL expressions that result in a numeric type. Like the other SQL Check Operators, only the first row returned by the given SQL queries will be used in the boolean evaluation. This is true for all of the parameters that can take a SQL query: `sql`, `min_threshold`, and `max_threshold`. This operator is great for use cases where a metric like a minimum, maximum, sum, or aggregate, is known to have to be between certain values.

Reviewer 3: ```suggestion
The `SQLThresholdCheckOperator` works similarly to the `SQLValueCheckOperator`. Instead of a single threshold, there is a min and max given, making this operator more dynamic in the type and scale of thresholding. The min and max thresholds can be numeric types or SQL expressions that result in a numeric type. Like the other SQL Check Operators, only the first row returned by the given SQL queries is used in the boolean evaluation. This is true for all of the parameters that can take a SQL query (`sql`, `min_threshold`, and `max_threshold`). This operator is great for use cases where a metric like a minimum, maximum, sum, or aggregate, has to be between certain values.
```

```python
SQLThresholdCheckOperator(
    task_id="check_threshold",
    sql="SELECT MAX(passenger_count)",
    min_threshold=1,
    max_threshold=8
)
```

> The above example ensures that the maximum passenger count for all rides is between 1 and 8 passengers, our assumed minimum and maximum number of potential riders for any type of taxicab.

Reviewer 3: Can you bring this description up to before the sample, like you do with the other examples?

## Conclusion

> After reading this guide, you should feel comfortable using any of the SQL Check Operators, understanding how each one works, and getting a sense of when each one would be useful in practice. With just these four operators, you have the foundation for a robust data quality suite right in your pipelines. If you are looking for more examples, or want to see how to use backend-specific operators like Redshift, BigQuery, or Snowflake, check out Astronomer's [data quality demo repository](https://github.com/astronomer/airflow-data-quality-demo/).

Reviewer 3: Awesome conclusion

- **[SQLIntervalCheckOperator](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/operators/sql/index.html#airflow.operators.sql.SQLIntervalCheckOperator)**: A time-based operator. Useful for checking current data against historical data
- **[SQLThresholdCheckOperator](https://airflow.apache.org/docs/apache-airflow/stable/_api/airflow/operators/sql/index.html#airflow.operators.sql.SQLThresholdCheckOperator)**: An operator with flexible upper and lower thresholds, where the threshold bounds may also be described as SQL queries that return a numeric value

## Examples

The following examples show you when and how to use each of the SQL Check Operators.

### Example 1 - `SQLCheckOperator`

The `SQLCheckOperator` returns a single row from a provided SQL query and checks to see if any of the returned values in that row are `False`. If any values are `False`, the task fails. This operator allows a great deal of flexibility in checking:

> - A specific, single column value

Reviewer 3: Because these complete a sentence when combined with the introductory clause, they can be punctuated.

- Part of or an entire row compared to a known set of values
- NULLs and types
- Specific calculations like `SUM`s or `AGG`s
- The results of any other function that can be written as a SQL query

The following code snippet shows you how to use the operator in a DAG:

```python
SQLCheckOperator(
    task_id="yellow_tripdata_row_quality_check",
    sql="row_quality_yellow_tripdata_check.sql",
    params={"pickup_datetime": "2021-01-01"},
)
```

> The `sql` argument can be either the complete SQL query as a string or, as in this example, a reference to a query in a local file (for Astronomer projects, this is in the `include/` directory). The `params` argument allows you to pass a dictionary of values to the SQL query, which can be accessed through the `params` keyword in the query. The `database` and `conn_id` arguments are left out of the example, so the default values will be used. The full code can be found in the [data quality demo repository](https://github.com/astronomer/airflow-data-quality-demo/).

Reviewer 3: ```suggestion
The `sql` argument can be either a complete SQL query as a string or, as in this example, a reference to a query in a local file (for Astronomer projects, this is in the `include/` directory). The `params` argument allows you to pass a dictionary of values to the SQL query, which can be accessed through the `params` keyword in the query. The `database` and `conn_id` arguments are left out of the example, so the default values will be used. The full code can be found in the [data quality demo repository](https://github.com/astronomer/airflow-data-quality-demo/).
```

The query used in the `sql` argument is:

```sql
SELECT vendor_id, pickup_datetime,
  CASE WHEN dropoff_datetime > pickup_datetime THEN 1 ELSE 0 END AS date_check,
  CASE WHEN passenger_count >= 0 THEN 1 ELSE 0 END AS passenger_count_check,
  CASE WHEN trip_distance >= 0 AND trip_distance <= 100 THEN 1 ELSE 0 END AS trip_distance_check,
  CASE WHEN ROUND((fare_amount + extra + mta_tax + tip_amount + improvement_surcharge + COALESCE(congestion_surcharge, 0)), 1) = ROUND(total_amount, 1) THEN 1
       WHEN ROUND(fare_amount + extra + mta_tax + tip_amount + improvement_surcharge, 1) = ROUND(total_amount, 1) THEN 1 ELSE 0 END AS fare_check
FROM yellow_tripdata
WHERE pickup_datetime IN (SELECT pickup_datetime FROM yellow_tripdata ORDER BY RANDOM() LIMIT 1)
```

> If instead of using a random `pickup_datetime`, we want to use a specific date to quality check a row, we can use the `params` passed into the operator like so:

Reviewer 3: ```suggestion
If we want to use a specific date to quality check a row instead of using a random `pickup_datetime`, we can use the `params` passed into the operator like so:
```
